optimum/habana/transformers/models/mistral/modeling_mistral.py

```diff
@@ -23,330 +23,537 @@
+        next_decoder_cache = () if not use_new_cache else None
+
+        for layer_idx, decoder_layer in enumerate(self.layers):
+            if layer_idx == len(self.layers) // 2:
+                htcore.mark_step()
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+
+            if self.gradient_checkpointing and self.training:
+                layer_outputs = self._gradient_checkpointing_func(
+                    decoder_layer.__call__,
+                    hidden_states,
+                    attention_mask,
+                    position_ids,
+                    None if past_key_values is None else past_key_values[layer_idx],
+                    output_attentions,
+                    use_cache,
+                    None,
+                )
+            else:
+                layer_outputs = decoder_layer(
+                    hidden_states,
+                    attention_mask=attention_mask,
+                    position_ids=position_ids,
+                    past_key_value=None if past_key_values is None else past_key_values[layer_idx],
+                    output_attentions=output_attentions,
+                    use_cache=use_cache,
+                    token_idx=token_idx,
+                    reuse_cache=reuse_cache,
+                    cache_idx=cache_idx,
+                    attn_softmax_bf16=attn_softmax_bf16,
+                )
+
+            hidden_states = layer_outputs[0]
+
+            if use_cache:
+                next_decoder_cache += (layer_outputs[2 if output_attentions else 1],)
+
+            if output_attentions:
+                all_self_attns += (layer_outputs[1],)
+
+        hidden_states = self.norm(hidden_states)
+
+        # add hidden states from the last decoder layer
+        if output_hidden_states:
+            all_hidden_states += (hidden_states,)
 
+        next_cache = None
         if use_cache:
-            next_decoder_cache += (layer_outputs[2 if output_attentions else 1],)
-
-        if output_attentions:
-            all_self_attns += (layer_outputs[1],)
+            next_cache = (
+                next_decoder_cache
+                if not use_new_cache
+                else (next_decoder_cache.to_legacy_cache() if use_legacy_cache else next_decoder_cache)
+            )
+        if not return_dict:
+            return tuple(v for v in [hidden_states, next_cache, all_hidden_states, all_self_attns] if v is not None)
+        return BaseModelOutputWithPast(
+            last_hidden_state=hidden_states,
+            past_key_values=next_cache,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attns,
+        )
 
-    hidden_states = self.norm(hidden_states)
 
-    # add hidden states from the last decoder layer
-    if output_hidden_states:
-        all_hidden_states += (hidden_states,)
+class GaudiMistralForCausalLM(MistralForCausalLM):
+    def allocate_kv_cache(self, batch_size, seq_len, _, __):
```

Comment:
Why do we need the extra two unused arguments?
```suggestion
    def allocate_kv_cache(self, batch_size, seq_len):
```

Reply:
the call is with 4 args here:

https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/generation/utils.py#L730